### PR TITLE
[Git Directory Explorer] - Easily navigate through directories containing Git repositories.

### PR DESCRIPTION
**Pull request by patriciomartinns**
## Description

Effortlessly navigate through your Git repositories and view commit history with Git Directory Explorer. Seamlessly open directories in your preferred code editor and streamline your development workflow.

## Screencast

Editar Projetos

View Commits

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

File: extensions/git-directory-explorer/src/git-directory-explorer.tsx [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+async function getSubdirectories(rootDir: string): Promise<string[]> {
+  const entries = await fs.promises.readdir(rootDir);
+  const directories = await Promise.all(
+    entries.map(async (entry) => {
+      const fullPath = path.join(rootDir, entry);
+      const entryStat = await fs.promises.stat(fullPath);
+      return entryStat.isDirectory() ? fullPath : null;
+    }),
+  );
```
Comment (pernielsentikaer) [inline extensions/git-directory-explorer/src/git-directory-explorer.tsx:39]: Could maybe be one function, something like (didn't test it)

```typescript
async function getGitSubdirectories(rootDir: string): Promise<string[]> {
  const entries = await fs.promises.readdir(rootDir);
  const gitDirectories = await Promise.all(
    entries.map(async (entry) => {
      const fullPath = path.join(rootDir, entry);
      const entryStat = await fs.promises.stat(fullPath);
      if (entryStat.isDirectory()) {
        const gitDir = path.join(fullPath, ".git");
        try {
          const stats = await fs.promises.stat(gitDir);
          if (stats.isDirectory()) {
            return fullPath;
          }
        } catch {
          // Not a git directory
        }
      }
      return null;
    })
  );
  return gitDirectories.filter(Boolean) as string[];
}
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi 👋 

Thanks for your contribution 💪 

I have now tested your extension, and I have some feedback ready for you:

- The extension suffers from [Empty State Flicker](<https://developers.raycast.com/basics/prepare-an-extension-for-store#empty-states>)

- The extension crashes if there is too many commits, in this case there is 26838 🫠

https://github.com/raycast/extensions/assets/38033107/946b7030-727b-4dbf-83c4-99ec05f66480

- Could you move your metadata (images placed in media) to a folder called metadata and change references in README

I'm looking forward to testing this extension again 🔥 

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

File: extensions/git-directory-explorer/src/git-directory-explorer.tsx [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+import fs from "fs";
+import path from "path";
+import os from "os";
+import simpleGit from "simple-git";
+
+interface Preferences {
+  directory: string;
+  codeEditor: Application;
+  terminalApp: Application;
+}
```
[the remark this answers is not in the document]
Comment (patriciomartinns) [reply extensions/git-directory-explorer/src/git-directory-explorer.tsx:12]: I understand, I hadn't considered that, I'll make the necessary corrections.

File: extensions/git-directory-explorer/src/git-directory-explorer.tsx (hunk 1 above)
Comment (patriciomartinns) [reply extensions/git-directory-explorer/src/git-directory-explorer.tsx:39]: Estou fazendo as validações com repositórios com numero de commits maiores, em breve lanço uma atualização para correção das falhas que foram reportadas por você.

**Comment by patriciomartinns** [comment]
I'm open to collaboration and who knows, it might help to improve this extension.

**Comment by mil3na** [comment]
Hi! 
Is this ready for another review?

**Comment by patriciomartinns** [comment]
Yes, I made the corrections that were requested

**Comment by patriciomartinns** [comment]
> Hi! Thanks again for your contribution! Now, it seems to work as expected. One little improvement I'd like to see is on the first load. There, I can see only a "Load More" button which might be confusing to the user. Do you mind adding some loading state?

Yes, I can add and request a new review, thanks for the recommendation.

**Comment by koinzhang** [comment]
Hi @patriciomartinns. I tried this extension and found that it's not quite what I expected.
1. you can only select a file and access the Git repository in the current folder; if the Git repository is in a subfolder, it won't be scanned out
2. This extension only seems to be able to view Git Commit information.

So I have a few suggestions
1. Wouldn't it be better to rename this extension to **Git Commits Explore**?
2. can we use _List.Item.Detail_ to display the first 30 commits directly on the List page, and then go to the secondary page if we want to see more?
3. Can we use the _Pagination_ function provided by Raycast to achieve the Commit information, which will be more natural and smooth?

**Pull request closed without merging**